### PR TITLE
[#1501] Add unit test for FeedbackContributionQuestionDetail

### PR DESCRIPTION
**Pull request by hhdqirui**
Part of #1501 

**Outline of Solution**

Add test for some of the methods that are not tested in FeedbackContributionQuestionDetail class.

### CODE REVIEW

File: src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java [hunk 1]
```diff
@@ -21,4 +32,173 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
...
+        feedbackQuestionAttributes.setGiverType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: recipient type can only be OWN_TEAM_MEMBERS_INCLUDING_SELF");
+        feedbackQuestionAttributes.setRecipientType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: invalid restrictions on visibility options");
```
Comment (rrtheonlyone) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java:197]: We can add more scenarios here - we only tested a little

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Really nice work 💯 I have put my comments below - just minor changes. 

Please also add tests for `getQuestionResultStatisticsJson` - that is the real challenge for this one!

File: src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java (hunk 1 above)
Comment (hhdqirui) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java:197]: Fixed

File: src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java [hunk 2]
```diff
@@ -21,4 +32,173 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
...
+            responses.add(details1);
+            FeedbackContributionResponseDetails details2 = new FeedbackContributionResponseDetails();
+            details2.setAnswer(Const.POINTS_NOT_SURE);
+            responses.add(details2);
+            FeedbackContributionResponseDetails details3 = new FeedbackContributionResponseDetails();
+            responses.add(details3);
+        }
+        assertTrue(feedbackContributionQuestionDetails.validateResponsesDetails(responses, 0).isEmpty());
+
+        ______TS("faliure: all answers of all responses are either not in range or are not multiple of 10");
```
[the remark this answers is not in the document]
Comment (hhdqirui) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java]: Fixed

**Comment by moziliar** [comment]
@hhdqirui can notify once the PR is ready for review

**Comment by rrtheonlyone** [comment]
@moziliar should be ready, please go ahead

**Comment by hhdqirui** [comment]
Hi @moziliar @rrtheonlyone . This PR is ready for review. Thank you!

File: src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java [hunk 3]
```diff
@@ -21,4 +37,568 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
...
+        feedbackQuestionAttributes.setGiverType(FeedbackParticipantType.STUDENTS);
+        feedbackQuestionAttributes.setRecipientType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: giver type is not STUDENT and recipient type is not OWN_TEAM_MEMBERS_INCLUDING_SELF");
+        feedbackQuestionAttributes.setGiverType(FeedbackParticipantType.SELF);
+        feedbackQuestionAttributes.setRecipientType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
```
Comment (moziliar) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java:570]: do you also wish to test the fallback `giverType` and `recipientType`? It seems to be part of the logic too.

File: src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java [hunk 4]
```diff
@@ -21,4 +37,568 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
...
+
+        ______TS("failure: giver type is not STUDENT and recipient type is not OWN_TEAM_MEMBERS_INCLUDING_SELF"
+                + " and invalid restrictions on visibility options");
+        feedbackQuestionAttributes.setGiverType(FeedbackParticipantType.SELF);
+        feedbackQuestionAttributes.setRecipientType(FeedbackParticipantType.SELF);
+        feedbackQuestionAttributes.setShowResponsesTo(Arrays.asList(FeedbackParticipantType.RECEIVER));
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_VISIBILITY_OPTIONS,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
```
Comment (moziliar) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java:601]: ditto: consider checking the fallback visibility types?

**Review by moziliar — CHANGES_REQUESTED** [review]
The test cases are quite comprehensive 👍👍 can't nitpick that much too

Please fix that one test which happened to pass because of coincidence (number of invalid options just nice coincide with the magic number 10 for loops; might want to set them differently too)

**Comment by moziliar** [comment]
Hi @hhdqirui any followup for this issue?

File: src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java (hunk 3 above)
Comment (hhdqirui) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java:570]: Thank you for your advice but I do not understand what does it mean by "fallback `giverType` and `recipientType`"? Can you explain further for me? Thank you!

Comment (moziliar) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java:570]: No problem! As you see in the method, the `giverType` is set to `STUDENTS` and `recipientType` is set to `OWN_TEAM_MEMBERS_INCLUDING_SELF` respectively upon detecting error. 
Do you intend to test this too?

Comment (hhdqirui) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java:570]: OK sure! Thank you for the explaination.

**Comment by hhdqirui** [comment]
> Hi @hhdqirui any followup for this issue?

@moziliar Hi the failed CI checks seems to be due to some reasons that are unrelated to this PR. May I know how should I proceed from here? Thank you!

**Comment by moziliar** [comment]
@hhdqirui no worries. When you face this kind of problem, you can rerun the CI in gh action.

**Comment by moziliar** [comment]
@hhdqirui are the reviews fixed? Can ping me after you are done:)

**Comment by hhdqirui** [comment]
> Hi @hhdqirui any followup for this issue?

@moziliar Yes, the reviews are fixed. It is ready for review. Thanks!

**Review by moziliar — APPROVED** [review]
Yup. The problems are fixed. I think the coverage is enough too for this class. Thanks!

**Review by rrtheonlyone — APPROVED** [review]
LGTM

**Pull request merged**